x/bep3/types/hash.go

```diff
@@ -0,0 +1,30 @@
+package types
+
+import (
+	"encoding/binary"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+)
+
+// CalculateRandomHash calculates the hash of a number and timestamp
+func CalculateRandomHash(randomNumber []byte, timestamp int64) []byte {
+	data := make([]byte, RandomNumberLength+Int64Size)
+	copy(data[:RandomNumberLength], randomNumber)
+	binary.BigEndian.PutUint64(data[RandomNumberLength:], uint64(timestamp))
+	return tmhash.Sum(data)
+}
+
+// CalculateSwapID calculates the hash of a RandomNumberHash, sdk.AccAddress, and string
+func CalculateSwapID(randomNumberHash string, sender sdk.AccAddress, senderOtherChain string) ([]byte, error) {
+	randomNumberHashBytes, err := HexEncodedStringToBytes(randomNumberHash)
+	if err != nil {
+		return []byte{}, err
+	}
+	senderOtherChain = strings.ToLower(senderOtherChain)
+	data := randomNumberHashBytes
+	data = append(data, []byte(sender)...)
+	data = append(data, []byte(senderOtherChain)...)
```

Comment:
Having `swapID` based on the string encoded address from the other chain could cause problems if they ever change the string encoding. Would it be possible to store other chain addresses as their underlying bytes instead?

Reply:
Yes, we could do that.

Looking into it I found that the [BEP3 documentation](https://github.com/binance-chain/BEPs/blob/master/BEP3.md#hash-timer-locked-transfer) indicates type bytes, yet the HTLTMsg in [Binance's go-sdk](https://github.com/binance-chain/go-sdk/blob/master/types/msg/msg-HTLT.go) defines them as strings:
```
RecipientOtherChain string           `json:"recipient_other_chain"`
SenderOtherChain    string           `json:"sender_other_chain"`
```

Reply:
Ah, found some additional info. We must implement BEP3's [common Executor interface](https://github.com/binance-chain/bep3-deputy/blob/master/common/executor.go), which defines these params in the HTLT function as strings:
```
HTLT(randomNumberHash common.Hash, timestamp int64, heightSpan int64, recipientAddr string, otherChainSenderAddr string, otherChainRecipientAddr string, outAmount *big.Int) (string, *Error)
```